python/kserve/kserve/protocol/rest/server.py

```diff
@@ -136,46 +137,67 @@ def __init__(self, http_port: int, sockets: List[socket.socket],
+                    "class": "logging.StreamHandler",
+                    "stream": "ext://sys.stdout",
+                },
+            },
+            "loggers": {
+                "uvicorn": {"handlers": ["default"], "level": "INFO"},
+                "uvicorn.error": {"level": "INFO"},
+                "uvicorn.access": {"handlers": ["access"], "level": "INFO"},
+            },
+        }
+
+        # If the log_config value is a string ending up with ".json"
+        # or ".yaml", it is interpreted as file path and the log configuration
+        # is loaded from disk.
+        if log_config_file:
+            log_config = log_config_file
+
         self.cfg = uvicorn.Config(
             app=app,
             host="0.0.0.0",
+            log_config=log_config,
             port=http_port,
-            log_config={
-                "version": 1,
-                "formatters": {
-                    "default": {
-                        "()": "uvicorn.logging.DefaultFormatter",
-                        "datefmt": DATE_FMT,
-                        "fmt": "%(asctime)s.%(msecs)03d %(name)s %(levelprefix)s %(message)s",
-                        "use_colors": None,
-                    },
-                    "access": {
-                        "()": "uvicorn.logging.AccessFormatter",
-                        "datefmt": DATE_FMT,
-                        "fmt": '%(asctime)s.%(msecs)03d %(name)s %(levelprefix)s %(client_addr)s %(process)s - '
-                               '"%(request_line)s" %(status_code)s',
-                        # noqa: E501
-                    },
-                },
-                "rest": {
-                    "default": {
-                        "formatter": "default",
-                        "class": "logging.StreamHandler",
-                        "stream": "ext://sys.stderr",
-                    },
-                    "access": {
-                        "formatter": "access",
-                        "class": "logging.StreamHandler",
-                        "stream": "ext://sys.stdout",
-                    },
-                },
-                "loggers": {
-                    "uvicorn": {"rest": ["default"], "level": "INFO"},
-                    "uvicorn.error": {"level": "INFO"},
-                    "uvicorn.access": {"rest": ["access"], "level": "INFO", "propagate": False},
-                },
-            }
         )
+
+        # More context in https://github.com/encode/uvicorn/pull/947
+        # At the time of writing the ASGI specs are not clear when it comes
+        # to change the access log format, and hence the Uvicorn upstream devs
+        # chose to create a custom middleware for this.
+        # The allowed log format is specified in https://github.com/Kludex/asgi-logger#usage
+        if access_log_format:
+            from asgi_logger import AccessLoggerMiddleware
+            logging.getLogger("uvicorn.access").handlers = []
+            app.add_middleware(
```

Comment:
How this is different from the `uvicorn.logging.AccessFormatter` config?

Reply:
My understanding from
* https://github.com/encode/uvicorn/blob/master/uvicorn/logging.py#L102
* https://github.com/encode/uvicorn/pull/947#issuecomment-1283006367
is that the access log formatter provided by uvicorn is not as powerful as other ones (like Gunicorn's). For example, in my use case I'd need to log request headers like User Agent to then be able to collect them in logstash and have a nice aggregation in Kibana, but other people may have different use cases. With the default access log formatter we get a limited format to use, meanwhile with `asgi-logger` we have way more possibilities available. It is sad that uvicorn didn't really decide what to do yet, but I suppose this is due to ASGI specs (see what's written in the github issue above). Note that if one wants to use and override `uvicorn.logging.AccessFormatter` they can via `log_config_file`.

Reply:
@elukey thanks for the explanation. We also have starlette metrics logging for printing the time spent on the endpoints, do you think we still need it with the uvicorn access logging?

Reply:
@yuzisun I'd keep it for the moment, so we can figure out with the community what's needed and what not. I didn't check but starlette shouldn't have its own access log right? Everything should go through uvicorn, but the precise relationship between the two is not 100% clear to me (if you want to shed some clarity I'd be very happy). If you want we can create a documentation page somewhere with all the details discussed in here, so folks in the community can refer to it and report use cases etc..

Reply:
Only two sources: uvicorn's access log (since `asgi-logger` overrides it, if used) and the startlette timer logging.

Reply:
ok, let's document all the logging arguments [here](https://kserve.github.io/website/0.10/modelserving/v1beta1/custom/custom_model/#arguments)?

Reply:
@yuzisun sure makes sense! Is it something that I can change in the KServe's repo or is it a sort-of-wiki/CMS that needs access to? (to understand if I have to modify the PR or not).

Reply:
I see https://github.com/kserve/website/blob/5d86aa0ead60b6e35c80f46e42866ab25e38aa1d/docs/modelserving/v1beta1/custom/custom_model/README.md, I'll file a separate pull request with the new info :)

Reply:
let’s create an issue in the website repo and submit the PR there

Reply:
@yuzisun done! https://github.com/kserve/website/pull/255

Reply:
mmm something must have changed, I tested the PR again and I see more logging than needed. Will do some research and report back.

Reply:
I do not see this field on the diff, where do you see that ?

Reply:
@yuzisun it is in the line:
`"uvicorn.access": {"rest": ["access"], "level": "INFO", "propagate": False`

I left out the `propagate` bit and I saw duplicated messages, and I noticed that the code to initialize the `AccessLoggerMiddleware` was not correct. Now I see the output as expected (namely as described above).

Reply:
And also third option, custom config file (note the `Custom-cfg-file` prefix in the access log formatter):

```
version: 1
disable_existing_loggers: False
formatters:
  default:
    (): 'uvicorn.logging.DefaultFormatter'
    fmt: '%(asctime)s %(levelprefix)-9s %(name)s -: %(message)s'
  access:
    (): 'uvicorn.logging.AccessFormatter'
    fmt: 'Custom-cfg-file - %(asctime)s %(levelprefix)-9s %(name)s -: %(client_addr)s - "%(request_line)s" %(status_code)s'
handlers:
  default:
    class: logging.StreamHandler
    formatter: default
    stream: ext://sys.stderr
  access:
    class: logging.StreamHandler
    formatter: access
    stream: ext://sys.stdout
loggers:
  uvicorn:
    level: INFO
    handlers:
      - default
  uvicorn.error:
    level: INFO
  uvicorn.access:
    level: INFO
    propagate: False
    handlers:
      - access
```

and logs:

```
2023-04-25 16:55:54.004 5476 root INFO [__call__():128] requestId: N.A., preprocess_ms: 0.015735626, explain_ms: 0, predict_ms: 0.00667572, 
postprocess_ms: 0.003099442                                                                                                                 
Custom-cfg-file - 2023-04-25 16:55:54,005 INFO:     uvicorn.access -: 127.0.0.1:35892 - "POST /v1/models/test%3Apredict HTTP/1.1" 200 OK    
2023-04-25 16:55:54.005 5476 root INFO [timing():48] kserve.io.kserve.protocol.rest.v1_endpoints.predict 0.0012352466583251953, ['http_statu
s:200', 'http_method:POST', 'time:wall']                                                                                                    
2023-04-25 16:55:54.005 5476 root INFO [timing():48] kserve.io.kserve.protocol.rest.v1_endpoints.predict 0.0012319999999998998, ['http_statu
s:200', 'http_method:POST', 'time:cpu'] 
```

Reply:
Almost! Sigh :(
When using asgi-logger I still see a little duplication:

```
127.0.0.1:41344 - - [25/Apr/2023:16:46:57 +0000] "POST /v1/models/test:predict HTTP/1.1" 200 - "-" "curl/7.74.0"
2023-04-25 16:46:57.906 5395 access INFO [log():67] 127.0.0.1:41344 - - [25/Apr/2023:16:46:57 +0000] "POST /v1/models/test:predict HTTP/1.1" 200 - "-" "curl/7.74.0"
```

In theory it shouldn't happen, maybe it is due how to asgi-logger sets the logger + propagation. Will investigate :)

Reply:
Ok should be fixed now, I added an extra setting for asgi-logger (to avoid propagation).

Reply:
Asked in https://github.com/Kludex/asgi-logger/issues/6 some guidance about what I have written, maybe there is another/better way.

Reply:
@yuzisun the change should be ready for a review, I checked and the output looks good now (no more repetitions).